### PR TITLE
[BEAM-6062] Spark runner do not show Beam metrics in Spark Web Interface

### PR DESCRIPTION
**Pull request by VaclavPlajt**
`MetricsContainerStepMap#toString` overwritten to show Beam metrics in Spark Web Interface.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

File: runners/spark/src/main/java/org/apache/beam/runners/spark/metrics/SparkMetricsContainerStepMap.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+/**
+ * Sole purpose of this class is to override {@link #toString()} of {@link MetricsContainerStepMap}
+ * in order to show meaningful metrics in Spark Web Interface.
+ */
+public class SparkMetricsContainerStepMap extends MetricsContainerStepMap {
+  private final transient SparkBeamMetric metricRenderer = new SparkBeamMetric();
+
+  @Override
+  public String toString() {
+    return metricRenderer.toString();
```
Comment (dmvk) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/metrics/SparkMetricsContainerStepMap.java]: can you please explain why this helps?  SparkBeamMetric does not have toString implementation

File: runners/spark/src/main/java/org/apache/beam/runners/spark/metrics/SparkMetricsContainerStepMap.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+package org.apache.beam.runners.spark.metrics;
+
+import org.apache.beam.runners.core.metrics.MetricsContainerStepMap;
+
+/**
+ * Sole purpose of this class is to override {@link #toString()} of {@link MetricsContainerStepMap}
+ * in order to show meaningful metrics in Spark Web Interface.
+ */
+public class SparkMetricsContainerStepMap extends MetricsContainerStepMap {
+  private final transient SparkBeamMetric metricRenderer = new SparkBeamMetric();
```
Comment (dmvk) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/metrics/SparkMetricsContainerStepMap.java]: why is this property transient? wouldn't this result in NPE when deserialized? (in toString method)

File: runners/spark/src/main/java/org/apache/beam/runners/spark/metrics/SparkMetricsContainerStepMap.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+  @Override
+  public boolean equals(Object o) {
+    return super.equals(o);
+  }
+
+  @Override
+  public int hashCode() {
+    return super.hashCode();
+  }
```
Comment (dmvk) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/metrics/SparkMetricsContainerStepMap.java]: unnecessary overrides

File: runners/spark/src/main/java/org/apache/beam/runners/spark/metrics/SparkMetricsContainerStepMap.java (hunk 1 above)
Comment (VaclavPlajt) [reply runners/spark/src/main/java/org/apache/beam/runners/spark/metrics/SparkMetricsContainerStepMap.java]: My fault, I just fixed it. `metricRenderer.renderAll()` is the right call to get all metrics in readable form.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/metrics/SparkMetricsContainerStepMap.java (hunk 3 above)
Comment (VaclavPlajt) [reply runners/spark/src/main/java/org/apache/beam/runners/spark/metrics/SparkMetricsContainerStepMap.java]: FindBugs complained about not overriding `equals()`. So I override `hashCode()` too, but you are right, It is unnecessary. Removed.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/metrics/SparkMetricsContainerStepMap.java (hunk 2 above)
Comment (VaclavPlajt) [reply runners/spark/src/main/java/org/apache/beam/runners/spark/metrics/SparkMetricsContainerStepMap.java]: Thanks

**Comment by dmvk** [comment]
Run Python PreCommit

**Comment by iemejia** [comment]
Run Spark ValidatesRunner

**Comment by iemejia** [comment]
Python build seems to be broken at the moment, that should not impact merging this PR since it does not touch python or the build files at all.

**Comment by VaclavPlajt** [comment]
Run Spark ValidatesRunner

**Pull request merged**